File: frontend/src/domain/audio/value-objects/sample-rate.ts
/**
 * Sample rate value object.
 */

export class SampleRate {
    private constructor(public readonly hertz: number) {
        if (hertz <= 0) {
            throw new Error('Sample rate must be positive');
        }
        if (hertz > 192000) {
            throw new Error('Sample rate cannot exceed 192kHz');
        }
    }

    static create(hertz: number): SampleRate {
        return new SampleRate(hertz);
    }

    static rate16k(): SampleRate {
        return new SampleRate(16000);
    }

    static rate22k(): SampleRate {
        return new SampleRate(22050);
    }

    static rate44k(): SampleRate {
        return new SampleRate(44100);
    }

    static rate48k(): SampleRate {
        return new SampleRate(48000);
    }

    get kilohertz(): number {
        return this.hertz / 1000;
    }

    equals(other: SampleRate): boolean {
        return this.hertz === other.hertz;
    }
}

File: backend/streamcraft/domain/audio/value_objects/sample_rate.py
"""Sample rate value object."""

from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class SampleRate:
    """Audio sample rate in Hz."""

    hertz: int

    def __post_init__(self) -> None:
        """Validate sample rate."""
        if self.hertz <= 0:
            raise ValueError("Sample rate must be positive")
        if self.hertz > 192000:
            raise ValueError("Sample rate cannot exceed 192kHz")

    @classmethod
    def rate_16k(cls) -> "SampleRate":
        """Create 16kHz sample rate."""
        return cls(hertz=16000)

    @classmethod
    def rate_22k(cls) -> "SampleRate":
        """Create 22.05kHz sample rate."""
        return cls(hertz=22050)

    @classmethod
    def rate_44k(cls) -> "SampleRate":
        """Create 44.1kHz sample rate."""
        return cls(hertz=44100)

    @classmethod
    def rate_48k(cls) -> "SampleRate":
        """Create 48kHz sample rate."""
        return cls(hertz=48000)

    @property
    def kilohertz(self) -> float:
        """Get sample rate in kHz."""
        return self.hertz / 1000.0
